# Ex 4.8 Grouping Data & Aggregating Variables

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path=r'C:\Users\jecka\2023-09-11 Achievement 4 Python'

In [3]:
# Import merged file
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_busy.pkl'))

In [4]:
df = ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 19)

In [6]:
df.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   
5                  7                   19.0       196.0                1.0   
6                  9                   20.0       196.0                1.0   
7                 14                   14.0       196.0                2.0   
8                 16                    0.0       196.0                4.0   
9                  8                   30.0       196.0                1.0   

   reordered product_name  aisle_id  department_id  prices _merge  \
0        0.0         Soda        77              7     9.0   both   
1        1.0         Soda        77              7     9.0   both   
2        1.0         Soda        77              7     9.0   both   
3        1.0         Soda        77              7     9.0   both   
4        1.0         Soda        77              7     9.0   both   
5        1.0         Soda        77              7     9.0   both   
6        1.0         Soda        77              7     9.0   both   
7        1.0         Soda        77              7     9.0   both   
8        1.0         Soda        77              7     9.0   both   
9        1.0         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy        Average orders  
1  Mid-range product  Regularly busy      Least busy        Average orders  
2  Mid-range product  Regularly busy      Least busy           Most orders  
3  Mid-range product      Least busy      Least busy        Average orders  
4  Mid-range product      Least busy      Least busy           Most orders  
5  Mid-range product  Regularly busy  Regularly busy        Average orders  
6  Mid-range product  Regularly busy     Busiest day           Most orders  
7  Mid-range product  Regularly busy     Busiest day           Most orders  
8  Mid-range product  Regularly busy     Busiest day           Most orders  
9  Mid-range product      Least busy      Least busy        Average orders

## Grouping Data with pandas

In [7]:
df.groupby('product_name')

## Aggregating Data with agg()

In [8]:
# Aggregate Department Id mean
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [9]:
# Alternate agg()
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [10]:
# Performing Multiple Agg
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

## Aggregating Data with transform()

In [11]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   

    order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                   8                    NaN       196.0                1.0   
1                   7                   15.0       196.0                1.0   
2                  12                   21.0       196.0                1.0   
3                   7                   29.0       196.0                1.0   
4                  15                   28.0       196.0                1.0   
5                   7                   19.0       196.0                1.0   
6                   9                   20.0       196.0                1.0   
7                  14                   14.0       196.0                2.0   
8                  16                    0.0       196.0                4.0   
9                   8                   30.0       196.0                1.0   
10                  9                    7.0       196.0                2.0   
11                 16                    8.0       196.0                6.0   
12                 11                    7.0       196.0                1.0   
13                  9                    7.0       196.0                1.0   
14                 10                   14.0       196.0                1.0   

    reordered product_name  aisle_id  department_id  prices _merge  \
0         0.0         Soda        77              7     9.0   both   
1         1.0         Soda        77              7     9.0   both   
2         1.0         Soda        77              7     9.0   both   
3         1.0         Soda        77              7     9.0   both   
4         1.0         Soda        77              7     9.0   both   
5         1.0         Soda        77              7     9.0   both   
6         1.0         Soda        77              7     9.0   both   
7         1.0         Soda        77              7     9.0   both   
8         1.0         Soda        77              7     9.0   both   
9         1.0         Soda        77              7     9.0   both   
10        0.0         Soda        77              7     9.0   both   
11        1.0         Soda        77              7     9.0   both   
12        1.0         Soda        77              7     9.0   both   
13        1.0         Soda        77              7     9.0   both   
14        1.0         Soda        77              7     9.0   both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy        Average orders   
1   Mid-range product  Regularly busy      Least busy        Average orders   
2   Mid-range product  Regularly busy      Least busy           Most orders   
3   Mid-range product      Least busy      Least busy        Average orders   
4   Mid-range product      Least busy      Least busy           Most orders   
5   Mid-range product  Regularly busy  Regularly busy  

In [13]:
# Remove limitation
pd.options.display.max_rows = None

In [ ]:
# Check datafram with more rows
ords_prods_merge.head(100)

##### Cleared output above for space

## Deriving Columns with loc()

In [15]:
# Adding Loyalty Flags
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [16]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [17]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
# Printing the frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [19]:
# mutliple column check
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

# Task 4.8

## Step 2

In [20]:
# Find the mean for entire dataframe
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3

#### How do the results for the entire dataframe differ from those of the subset?

In [21]:
# Subset Avg to compare
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

Because the subset was limited, the numbers differ from the merged dataframe (ords_prods). Some of them are not off by much (Ex: 7 **Subset**: 17.47; **ords_prods**: 17.23). 
The average in the subsets are larger with the exception of 17 (**Subset**: 11.29; **ords_prods**: 15.69)

## Step 4

### Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

#### This was done with the rest of the lesson.
#### See "Deriving Columns with loc()".

## Step 5

### Difference between the spending habits 
###  Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer)

In [22]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

The min and max prices are the same for all three types of customers. The Mean for New Customers is the highest, followed by regular customers and loyal customers have the lowest mean.


## Step 6

### Create a spending flag for each user based on the average price across all their orders


In [23]:
# Split the data into groups based on the 'user_id' column.
# Apply the transform() function on the 'prices' to get the avg price
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [24]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] >=10, 'spender_type'] = 'High spender'

In [25]:
ords_prods_merge.loc[ords_prods_merge['avg_price'] <10, 'spender_type'] = 'Low spender'

In [26]:
ords_prods_merge['spender_type'].value_counts(dropna = False)

spender_type
Low spender     31770614
High spender      634245
Name: count, dtype: int64

In [27]:
ords_prods_merge.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  \
0                  8                    NaN       196.0                1.0   
1                  7                   15.0       196.0                1.0   
2                 12                   21.0       196.0                1.0   
3                  7                   29.0       196.0                1.0   
4                 15                   28.0       196.0                1.0   
5                  7                   19.0       196.0                1.0   
6                  9                   20.0       196.0                1.0   
7                 14                   14.0       196.0                2.0   
8                 16                    0.0       196.0                4.0   
9                  8                   30.0       196.0                1.0   

   reordered  ... prices  _merge    price_range_loc     busiest_day  \
0        0.0  ...    9.0    both  Mid-range product  Regularly busy   
1        1.0  ...    9.0    both  Mid-range product  Regularly busy   
2        1.0  ...    9.0    both  Mid-range product  Regularly busy   
3        1.0  ...    9.0    both  Mid-range product      Least busy   
4        1.0  ...    9.0    both  Mid-range product      Least busy   
5        1.0  ...    9.0    both  Mid-range product  Regularly busy   
6        1.0  ...    9.0    both  Mid-range product  Regularly busy   
7        1.0  ...    9.0    both  Mid-range product  Regularly busy   
8        1.0  ...    9.0    both  Mid-range product  Regularly busy   
9        1.0  ...    9.0    both  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag avg_price  \
0  Regularly busy        Average orders        10  New customer  6.367797   
1      Least busy        Average orders        10  New customer  6.367797   
2      Least busy           Most orders        10  New customer  6.367797   
3      Least busy        Average orders        10  New customer  6.367797   
4      Least busy           Most orders        10  New customer  6.367797   
5  Regularly busy        Average orders        10  New customer  6.367797   
6     Busiest day           Most orders        10  New customer  6.367797   
7     Busiest day           Most orders        10  New customer  6.367797   
8     Busiest day           Most orders        10  New customer  6.367797   
9      Least busy        Average orders        10  New customer  6.367797   

   spender_type  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  
5   Low spender  
6   Low spender  
7   Low spender  
8   Low spender  
9   Low spender  

[10 rows x 23 columns]

## Step 7

### Determine frequent versus non-frequent customers.

In [28]:
# Split the data into groups based on the 'user_id' column.
# Apply the transform() function on the 'days_since_last_order' to get the median
ords_prods_merge['past_orders_median'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [29]:
ords_prods_merge.loc[ords_prods_merge['past_orders_median'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [30]:
ords_prods_merge.loc[(ords_prods_merge['past_orders_median'] <= 20) & (ords_prods_merge['past_orders_median'] > 10), 'frequency_flag'] = 'Regular customer'

In [31]:
ords_prods_merge.loc[ords_prods_merge['past_orders_median'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [32]:
# Printing the frequency
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [33]:
# Checking multiple columns
ords_prods_merge[['user_id', 'past_orders_median', 'frequency_flag']].head(20)

user_id  past_orders_median         frequency_flag
0         1                20.5  Non-frequent customer
1         1                20.5  Non-frequent customer
2         1                20.5  Non-frequent customer
3         1                20.5  Non-frequent customer
4         1                20.5  Non-frequent customer
5         1                20.5  Non-frequent customer
6         1                20.5  Non-frequent customer
7         1                20.5  Non-frequent customer
8         1                20.5  Non-frequent customer
9         1                20.5  Non-frequent customer
10       15                10.0      Frequent customer
11       15                10.0      Frequent customer
12       15                10.0      Frequent customer
13       15                10.0      Frequent customer
14       15                10.0      Frequent customer
15       19                 6.0      Frequent customer
16       19                 6.0      Frequent customer
17       19                 6.0      Frequent customer
18       21                 7.5      Frequent customer
19       31                 6.0      Frequent customer

In [34]:
ords_prods_merge[['user_id', 'past_orders_median', 'frequency_flag']].tail(10)

user_id  past_orders_median     frequency_flag
32404849   202557                 8.0  Frequent customer
32404850   202557                 8.0  Frequent customer
32404851   202557                 8.0  Frequent customer
32404852   202557                 8.0  Frequent customer
32404853   202557                 8.0  Frequent customer
32404854   202557                 8.0  Frequent customer
32404855   202557                 8.0  Frequent customer
32404856   203436                15.0   Regular customer
32404857   203436                15.0   Regular customer
32404858   205420                13.0   Regular customer

In [36]:
# Export df to pkl
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_group_transform.pkl'))